src/core/ext/filters/client_channel/resolver/dns/event_engine/event_engine_client_channel_resolver.cc

```diff
@@ -0,0 +1,535 @@
+                                                        *balancer_addresses_);
+  }
+}
+
+void EventEngineDNSRequestWrapper::MaybePopulateServiceConfigLocked(
+    grpc_core::Resolver::Result& result) {
+  // TODO(roth): We should not ignore errors here, but instead have users of the
+  // resolver handle them appropriately. Some test/cpp/naming tests fail.
+  if ((addresses_.ok() || balancer_addresses_.ok()) &&
+      service_config_json_.ok() && service_config_json_->empty()) {
+    result.service_config = service_config_json_.status();
+  }
+  if (service_config_json_.ok() && !service_config_json_->empty()) {
+    auto service_config = ChooseServiceConfig(*service_config_json_);
+    if (!service_config.ok()) {
+      result.service_config = absl::UnavailableError(
+          absl::StrCat("failed to parse service config: ",
+                       grpc_core::StatusToString(service_config.status())));
+    } else if (!service_config->empty()) {
+      GRPC_EVENT_ENGINE_DNS_TRACE(
+          "resolver:%p selected service config choice: %s", this,
+          service_config->c_str());
+      result.service_config = grpc_core::ServiceConfigImpl::Create(
+          resolver_->channel_args(), *service_config);
+      if (!result.service_config.ok()) {
+        result.service_config = absl::UnavailableError(
+            absl::StrCat("failed to parse service config: ",
+                         result.service_config.status().message()));
+      }
+    }
+  }
+}
+
+void EventEngineDNSRequestWrapper::NormalizeEmptyResultsToErrorsLocked() {
+  if (addresses_.ok() && addresses_->empty()) {
+    addresses_ = absl::NotFoundError("No addresses found.");
+  }
+  if (balancer_addresses_.ok() && balancer_addresses_->empty()) {
+    balancer_addresses_ = absl::NotFoundError("No addresses found.");
+  }
+}
+
+absl::optional<grpc_core::Resolver::Result>
+EventEngineDNSRequestWrapper::OnResolvedLocked() {
+  if (orphaned_) return absl::nullopt;
+  // Wait for all requested queries to return.
+  if (hostname_handle_.has_value() || srv_handle_.has_value() ||
+      txt_handle_.has_value() ||
+      number_of_balancer_hostnames_resolved_ !=
+          balancer_hostname_handles_.size()) {
+    GRPC_EVENT_ENGINE_DNS_TRACE(
+        "resolver:%p OnResolved() waiting for results (hostname: %s, "
+        "srv: %s, "
+        "txt: %s, "
+        "balancer addresses: %zu/%zu complete",
+        this, hostname_handle_.has_value() ? "waiting" : "done",
+        srv_handle_.has_value() ? "waiting" : "done",
+        txt_handle_.has_value() ? "waiting" : "done",
+        number_of_balancer_hostnames_resolved_,
+        balancer_hostname_handles_.size());
+    return absl::nullopt;
+  }
+  GRPC_EVENT_ENGINE_DNS_TRACE("resolver:%p OnResolvedLocked() proceeding",
+                              this);
+  grpc_core::Resolver::Result result;
+  result.args = resolver_->channel_args();
+  NormalizeEmptyResultsToErrorsLocked();
+  MaybePopulateServiceConfigLocked(result);
```

Comment:
As mentioned above, I think we need to have all of the logic for handling address failures in one place, which is right here.  The logic should be something like this:

```
// If both addresses and balancer addresses failed, return an error for
// both addresses and service config.
if (!addresses_.ok() && !balancer_addresses_.ok()) {
  absl::Status status = GetResolutionFailureErrorMessageLocked();
  result.addresses = status;
  result.service_confg = status;
  return result;
}
// We have at least one of addresses or balancer addresses, so we're
// going to return a non-error for addresses.
result.addresses.emplace();
MaybePopulateAddressesLocked(result);
MaybePopulateBalancerAddressesLocked(result);
MaybePopulateServiceConfigLocked(result);
return result;
```